velox/exec/WindowPartition.cpp

```diff
@@ -130,47 +192,104 @@ bool WindowPartition::compareRowsWithSortKeys(const char* lhs, const char* rhs)
   return false;
 }
 
+vector_size_t WindowPartition::findPeerRowEndIndex(
+    vector_size_t currentStart,
+    vector_size_t lastPartitionRow,
+    const std::function<bool(const char*, const char*, RowContainer* data)>&
+        peerCompare) {
+  auto peerEnd = currentStart;
+  while (peerEnd <= lastPartitionRow) {
+    if (peerCompare(
+            partition_[currentStart - startRow_],
+            partition_[peerEnd - startRow_],
+            data_)) {
+      break;
+    }
+    ++peerEnd;
+  }
+  return peerEnd;
+}
+
 std::pair<vector_size_t, vector_size_t> WindowPartition::computePeerBuffers(
     vector_size_t start,
     vector_size_t end,
     vector_size_t prevPeerStart,
     vector_size_t prevPeerEnd,
     vector_size_t* rawPeerStarts,
-    vector_size_t* rawPeerEnds) const {
-  const auto peerCompare = [&](const char* lhs, const char* rhs) -> bool {
-    return compareRowsWithSortKeys(lhs, rhs);
+    vector_size_t* rawPeerEnds) {
+  const auto peerCompare =
+      [&](const char* lhs, const char* rhs, RowContainer* data) -> bool {
+    return compareRowsWithSortKeys(lhs, rhs, data);
   };
 
-  VELOX_CHECK_LE(end, numRows());
+  VELOX_CHECK_LE(end, numRows() + startRow_);
 
-  const auto lastPartitionRow = numRows() - 1;
+  auto lastPartitionRow = numRows() + startRow_ - 1;
   auto peerStart = prevPeerStart;
   auto peerEnd = prevPeerEnd;
-  for (auto i = start, j = 0; i < end; ++i, ++j) {
-    // When traversing input partition rows, the peers are the rows with the
-    // same values for the ORDER BY clause. These rows are equal in some ways
-    // and affect the results of ranking functions. This logic exploits the fact
-    // that all rows between the peerStart and peerEnd have the same values for
-    // rawPeerStarts and rawPeerEnds. So we can compute them just once and reuse
-    // across the rows in that peer interval. Note: peerStart and peerEnd can be
-    // maintained across getOutput calls. Hence, they are returned to the
-    // caller.
+
+  auto nextStart = start;
+
+  if (partial_ && start > 0) {
+    VELOX_CHECK(previousRow_);
+
+    // rowContainer is used to compare whether the current first row is in same
+    // peer with the last row of previous batch.
+    auto rowContainer =
+        std::make_unique<RowContainer>(data_->keyTypes(), pool_);
+    auto serialized =
+        BaseVector::create<FlatVector<StringView>>(VARBINARY(), 1, pool_);
+    data_->extractSerializedRows(
+        folly::Range(rows_.data() + (start - startRow_), 1), serialized);
+
+    auto firstRow = rowContainer->newRow();
+    rowContainer->storeSerializedRow(*serialized, 0, firstRow);
+    auto secondRow = rowContainer->newRow();
+    rowContainer->storeSerializedRow(*previousRow_, 0, secondRow);
+
+    auto peerGroup = peerCompare(secondRow, firstRow, rowContainer.get());
```

Comment:
@xiaoxmeng Got the following exception when using the previous row to compare peer group. And this issue is not stable.

```
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007f4b695d8467, pid=3989197, tid=0x00007f4d62df9700
#
# JRE version: OpenJDK Runtime Environment (8.0_422-b05) (build 1.8.0_422-8u422-b05-1~20.04-b05)
# Java VM: OpenJDK 64-Bit Server VM (25.422-b05 mixed mode linux-amd64 compressed oops)
# Problematic frame:
# C  [libvelox.so+0x8bf9467]  xsimd::batch<signed char, xsimd::fma3<xsimd::avx2> > xsimd::kernel::load_unaligned<xsimd::fma3<xsimd::avx2>, signed char, void>(signed char const*, xsimd::kernel::convert<signed char>, xsimd::avx const&)+0x3e
#
# Failed to write core dump. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /mnt/DP_disk3/jk/projects/gluten/tools/gluten-it/hs_err_pid3989197.log
#
# If you would like to submit a bug report, please visit:
#   http://bugreport.java.com/bugreport/crash.jsp
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
#
```

And the call stack is as following.

```
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
C  [libvelox.so+0x8bf9467]  xsimd::batch<signed char, xsimd::fma3<xsimd::avx2> > xsimd::kernel::load_unaligned<xsimd::fma3<xsimd::avx2>, signed char, void>(signed char const*, xsimd::kernel::convert<signed char>, xsimd::avx const&)+0x3e
C  [libvelox.so+0x8bf75aa]  xsimd::batch<signed char, xsimd::fma3<xsimd::avx2> > xsimd::batch<signed char, xsimd::fma3<xsimd::avx2> >::load_unaligned<signed char>(signed char const*)+0x46
C  [libvelox.so+0x8bf4751]  facebook::velox::simd::detail::CopyWord<xsimd::batch<signed char, xsimd::fma3<xsimd::avx2> >, xsimd::fma3<xsimd::avx2> >::apply(void*, void const*)+0x37
C  [libvelox.so+0x8bf0ea4]  bool facebook::velox::simd::detail::copyNextWord<xsimd::batch<signed char, xsimd::fma3<xsimd::avx2> >, xsimd::fma3<xsimd::avx2> >(void*&, void const*&, int&)+0x3f
C  [libvelox.so+0x8bed938]  void facebook::velox::simd::memcpy<xsimd::fma3<xsimd::avx2> >(void*, void const*, int, xsimd::fma3<xsimd::avx2> const&)+0x4a
C  [libvelox.so+0x8e702da]  facebook::velox::ByteOutputStream::appendStringView(std::basic_string_view<char, std::char_traits<char> >)+0xca
C  [libvelox.so+0x8e70208]  facebook::velox::ByteOutputStream::appendStringView(facebook::velox::StringView)+0x46
C  [libvelox.so+0xb999520]  facebook::velox::HashStringAllocator::copyMultipartNoInline(facebook::velox::StringView const&, char*, int)+0x106
C  [libvelox.so+0xb847346]  facebook::velox::HashStringAllocator::copyMultipart(facebook::velox::StringView const&, char*, int)+0x64
C  [libvelox.so+0xb840fcd]  facebook::velox::exec::RowContainer::storeVariableSizeAt(char const*, char*, unsigned int)+0x11b
C  [libvelox.so+0xb841a5e]  facebook::velox::exec::RowContainer::storeSerializedRow(facebook::velox::FlatVector<facebook::velox::StringView> const&, int, char*)+0x24c
C  [libvelox.so+0xb975bde]  facebook::velox::exec::WindowPartition::computePeerBuffers(int, int, int, int, int*, int*)+0x33e
C  [libvelox.so+0xe8e9481]  facebook::velox::exec::Window::computePeerAndFrameBuffers(int, int)+0x2f9
C  [libvelox.so+0xe8e9844]  facebook::velox::exec::Window::callApplyForPartitionRows(int, int, int, std::shared_ptr<facebook::velox::RowVector> const&)+0x3a
C  [libvelox.so+0xe8e9ae3]  facebook::velox::exec::Window::callApplyLoop(int, std::shared_ptr<facebook::velox::RowVector> const&)+0xe1
C  [libvelox.so+0xe8e9da8]  facebook::velox::exec::Window::getOutput()+0x1d6
C  [libvelox.so+0xe7cc3bf]  facebook::velox::exec::Driver::runInternal(std::shared_ptr<facebook::velox::exec::Driver>&, std::shared_ptr<facebook::velox::exec::BlockingState>&, std::shared_ptr<facebook::velox::RowVector>&)+0xf8b
C  [libvelox.so+0xe7ca170]  facebook::velox::exec::Driver::next(folly::SemiFuture<folly::Unit>*)+0xf0
C  [libvelox.so+0xb8b3770]  facebook::velox::exec::Task::next(folly::SemiFuture<folly::Unit>*)+0x750
C  [libvelox.so+0x8985edb]  gluten::WholeStageResultIterator::next()+0x89
C  [libgluten.so+0x1b058a2]  gluten::ResultIterator::getNext()+0x5a
C  [libgluten.so+0x1b0567c]  gluten::ResultIterator::hasNext()+0x28

```